### PR TITLE
[WIP][guilib/pictures] Add an ffmpeg based image loader

### PR DESCRIPTION
**Pull request by ace20022**
Taken from https://github.com/notspiff/kodi-cmake/commit/b93fa47c2f9fa8becd32edbcab7e0ac3c5d3e307

### CODE REVIEW

**Comment by it-solutions-zehaczek** [comment]
awesome! =)

**Comment by mkortstiege** [comment]
@ace20022 please pick and squash https://github.com/mkortstiege/xbmc/commit/64a8d9de1a97a2d4e3aa0b4a98ecc2a66aed520b. This contains the xcode sync to make it build on ios/osx

File: xbmc/guilib/FFmpegImage.cpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
+/*
+ *      Copyright (C) 2012-2013 Team XBMC
+ *      http://xbmc.org
```
Comment (hudokkow) [inline xbmc/guilib/FFmpegImage.cpp]: spiff shouldn't mind... :wink: 
Team Kodi
kodi.tv

Same on other files

File: xbmc/guilib/FFmpegImage.cpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+CFFmpegImage::~CFFmpegImage()
+{
+}
+
+bool CFFmpegImage::LoadImageFromMemory(unsigned char* buffer, unsigned int bufSize,
+                                      unsigned int width, unsigned int height)
+{
+  uint8_t* fbuffer = (uint8_t*)av_malloc(32768);
```
Comment (ace20022) [inline xbmc/guilib/FFmpegImage.cpp]: Use FFMPEG_FILE_BUFFER_SIZE

**Comment by ace20022** [comment]
jenkins build this please

@notspiff @FernetMenta @fritsch Could you please review?
There seems to be a heap problem with this file http://thundercloud.net/infoave/images/2014/tropical-sunrise.webp , but I had no time to investigate it.

**Comment by ace20022** [comment]
thx @notspiff for the fixes and thx @fritsch for the review, will look into it in more detail asasp.

**Comment by ace20022** [comment]
>This will go wrong, if src's pitch is smaller than yours, e.g. src's pitch is 8 aligned and yours is 16 aligned - then you will overread data, access data that is not yours.

Imo we have to change the  interface to make it right, since we don't know the hight and therefore the size of dst.

**Comment by ace20022** [comment]
> care to use a namespace for these definitions?

if you post the code np

**Comment by fritsch** [comment]
You know the size and the pitch and as the code calls:
```
  if(pImage != NULL && pImage->LoadImageFromMemory(buffer, bufSize, width, height))
  {
    if (pImage->Width() > 0 && pImage->Height() > 0)
    {
      Allocate(pImage->Width(), pImage->Height(), XB_FMT_A8R8G8B8);
      if (pImage->Decode(m_pixels, GetPitch(), XB_FMT_A8R8G8B8)
...
```

So it asks on your pImage for the Width and Height, which you also know internally. So for computing the dst pitch, all information is known right?

**Comment by ace20022** [comment]
You have to look at allocate.

**Comment by fritsch** [comment]
Yeah, that's what I did. Allocate uses your m_height and m_width. I added a code snipped at the function we started the discussion, simple idea: only read the min of the pitches available ... which will work in any case.

**Comment by ace20022** [comment]
> Yeah, that's what I did.

We have:
```m_pixels = new unsigned char[GetPitch() * GetRows()]```
Therefore we allocate GetPitch() * GetRows() bytes. That is
```
GetPitch() * GetRows() =
GetPitch(m_textureWidth) * GetRows(m_textureHeight) =
```
```m_textureWidth``` and ```m_textureHeight``` might be lesser than width and height, see clamp and the code before.

**Comment by fritsch** [comment]
Yes - that's a problem. Btw. for all implementations. On hw we support this is a theoretical minimum of 8k x 8k. So if you would create a 32k x 32k image all our image implementations would most likely segfault :-)

**Comment by ace20022** [comment]
Such posts are really annoying

**Comment by fritsch** [comment]
@ace20022 why? It was a question and I asked if you see it the same way?

**Comment by ace20022** [comment]
Update: I'm working on a fix for gifs for 16...

**Pull request closed without merging**